digitalocean/partnerinterconnectattachment/resource_partner_interconnect_attachment.go

```diff
@@ -0,0 +1,289 @@
+							Optional: true,
+						},
+						"auth_key": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(5 * time.Minute),
+			Delete: schema.DefaultTimeout(2 * time.Minute),
+		},
+	}
+}
+
+func resourceDigitalOceanPartnerInterconnectAttachmentCreate(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	client := meta.(*config.CombinedConfig).GodoClient()
+
+	name := d.Get("name").(string)
+	connectionBandwidthInMbps := d.Get("connection_bandwidth_in_mbps").(int)
+	region := d.Get("region").(string)
+	naasProvider := d.Get("naas_provider").(string)
+	vpcIDs := d.Get("vpc_ids").(*schema.Set).List()
+
+	vpcIDsString := make([]string, len(vpcIDs))
+	for i, v := range vpcIDs {
+		vpcIDsString[i] = v.(string)
+	}
+
+	partnerInterconnectAttachmentRequest := &godo.PartnerInterconnectAttachmentCreateRequest{
+		Name:                      name,
+		ConnectionBandwidthInMbps: connectionBandwidthInMbps,
+		Region:                    region,
+		NaaSProvider:              naasProvider,
+		VPCIDs:                    vpcIDsString,
+	}
+
+	if bgpRaw, ok := d.GetOk("bgp"); ok {
+		bgpList := bgpRaw.([]interface{})
+		if len(bgpList) > 0 {
+			bgpConfig := bgpList[0].(map[string]interface{})
+			bgp := godo.BGP{
+				LocalRouterIP: bgpConfig["local_router_ip"].(string),
+				PeerASN:       bgpConfig["peer_router_asn"].(int),
+				PeerRouterIP:  bgpConfig["peer_router_ip"].(string),
+				AuthKey:       bgpConfig["auth_key"].(string),
+			}
+			partnerInterconnectAttachmentRequest.BGP = bgp
+		}
+	}
+
+	log.Printf("[DEBUG] Partner Interconnect Attachment create request: %#v", partnerInterconnectAttachmentRequest)
+
+	err := retry.RetryContext(ctx, d.Timeout(schema.TimeoutCreate), func() *retry.RetryError {
+		partnerInterconnectAttachment, _, err := client.PartnerInterconnectAttachments.Create(context.Background(), partnerInterconnectAttachmentRequest)
+		if err != nil {
+			return retry.RetryableError(fmt.Errorf("error creating Partner Interconnect Attachment: %s", err))
+		}
+
+		d.SetId(partnerInterconnectAttachment.ID)
+
+		log.Printf("[DEBUG] Waiting for Partner Interconnect Attachment (%s) to become active", d.Get("name"))
+		stateConf := &retry.StateChangeConf{
+			Delay:      5 * time.Second,
+			Pending:    []string{"CREATING"},
+			Target:     []string{"CREATED"},
+			Refresh:    partnerInterconnectAttachmentStateRefreshFunc(client, d.Id()),
+			Timeout:    2 * time.Minute,
+			MinTimeout: 5 * time.Second,
+		}
+		if _, err := stateConf.WaitForStateContext(ctx); err != nil {
+			return retry.NonRetryableError(fmt.Errorf("error waiting for Partner Interconnect Attachment (%s) to become active: %s", d.Get("name"), err))
+		}
+		return nil
+	})
+
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	log.Printf("[INFO] Partner Interconnect Attachment created, ID: %s", d.Id())
+
+	return nil
```

Comment:
After creating the resource, you never set any of its attributes to state. One common pattern is to return the read method here to set everything. Though that does add an additional HTTP request. For example:

https://github.com/digitalocean/terraform-provider-digitalocean/blob/da119aebd61f90fbdc3f5d22db5eb1de63f4709f/digitalocean/vpc/resource_vpc.go#L116